monai/networks/nets/dints.py

```diff
@@ -0,0 +1,556 @@
+
+
+class _FactorizedReduceBlockWithMemory(FactorizedReduceBlock):
+    """
+    Down-sampling the feature by 2 using stride.
+    """
+
+    def __init__(self, in_channel: int, out_channel: int):
+        super().__init__(in_channel, out_channel)
+
+        # multiply by 8 to comply with cell output size (see net.get_memory_usage)
+        self.memory = (1 + out_channel / in_channel / 8 * 3) * 8 * in_channel / out_channel
+
+
+# Define Operation Set
+OPS = {
+    "skip_connect": lambda c: _IdentityWithMemory(),
+    "conv_3x3x3": lambda c: _ReLUConvBNBlockWithMemory(c, c, 3, padding=1),
+    "conv_3x3x1": lambda c: _P3DReLUConvBNBlockWithMemory(c, c, 3, padding=1, p3dmode=0),
+    "conv_3x1x3": lambda c: _P3DReLUConvBNBlockWithMemory(c, c, 3, padding=1, p3dmode=1),
+    "conv_1x3x3": lambda c: _P3DReLUConvBNBlockWithMemory(c, c, 3, padding=1, p3dmode=2),
+}
+
+
+class MixedOp(nn.Module):
+    def __init__(self, c, code_c=None):
+        super().__init__()
+        self._ops = nn.ModuleList()
+        if code_c is None:
+            code_c = np.ones(len(OPS))
+        for idx, _ in enumerate(OPS.keys()):
+            if idx < len(code_c):
+                if code_c[idx] == 0:
+                    op = None
+                else:
+                    op = OPS[_](c)
+                self._ops.append(op)
+
+    def forward(self, x, ops=None, weight: bool = None):
+        pos = (ops == 1).nonzero()
+        result = 0
+        for _ in pos:
+            result += self._ops[_.item()](x) * ops[_.item()] * weight[_.item()]
+        return result
+
+
+class Cell(nn.Module):
+    """
+    The basic class for cell operation
+    Args:
+        c_prev: input channel number
+        C: output channel number
+        rate: resolution change rate. -1 for 2x downsample, 1 for 2x upsample
+              0 for no change of resolution
+        code_c: cell operation code
+    """
+
+    def __init__(self, c_prev, c, rate: int, code_c: bool = None):
+        super().__init__()
+        self.c_out = c
+        if rate == -1:  # downsample
+            self.preprocess = _FactorizedReduceBlockWithMemory(c_prev, c)
+        elif rate == 1:  # upsample
+            self.preprocess = _FactorizedIncreaseBlockWithMemory(c_prev, c)
+        else:
+            if c_prev == c:
+                self.preprocess = _IdentityWithMemory()
+            else:
+                self.preprocess = _ReLUConvBNBlockWithMemory(c_prev, c, 1, 0)
+        self.op = MixedOp(c, code_c)
+
+    def forward(self, s, ops, weight):
+        s = self.preprocess(s)
+        s = self.op(s, ops, weight)
+        return s
+
+
+class DiNTS(nn.Module):
+    def __init__(
+        self,
+        in_channels: int,
+        num_classes: int,
+        cell=Cell,
+        cell_ops: int = 5,
+        channel_mul: float = 1.0,
+        num_blocks: int = 6,
+        num_depths: int = 3,
+        use_stem: bool = False,
+        code: list = None,
+    ):
+        """
+        Initialize NAS network search space
```

Comment:
would be great to have a short paragraph to describe the network including the paper references.